### PR TITLE
Rewrite ProtectedCommand and AsyncProperty. Add unit tests.

### PR DESCRIPTION
**Pull request by ILMTitan**
This came about while trying to fix a transient unit test failure: https://ci.appveyor.com/project/ILMTitan/google-cloud-visualstudio/build/2.0.0-dev.266/tests

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        /// <summary>Defines the method to be called when the command is invoked.</summary>
+        /// <param name="argument">Data used by the command.  If the command does not require data to be passed, this object can be set to null.</param>
+        public override void Execute(object argument)
+        {
+            Task actionTask = _asyncAction();
+            ErrorHandlerUtils.HandleExceptionsAsync(() => actionTask);
+            LatestExecution = new AsyncProperty(actionTask);
+        }
+
+        protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
```
Comment (redborian) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs]: Why virtual? This is very rarely overridden.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/CloudExplorerViewModel.cs [hunk 2]
```diff
@@ -427,7 +428,7 @@ private void InvalidateAccountDependentDataSources()
         /// </summary>
         private void NotifySourcesOfUpdatedAccountOrProject()
         {
-            foreach (var source in _sources)
+            foreach (ICloudExplorerSource<ISourceRootViewModelBase> source in _sources)
```
Comment (redborian) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/CloudExplorerViewModel.cs]: Is this for better readability? Or something more?

File: GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            LatestExecution = new AsyncProperty(Task.CompletedTask);
+        }
+
+        /// <summary>Defines the method to be called when the command is invoked.</summary>
+        /// <param name="argument">Data used by the command.  If the command does not require data to be passed, this object can be set to null.</param>
+        public override void Execute(object argument)
+        {
+            Task actionTask = _asyncAction();
+            ErrorHandlerUtils.HandleExceptionsAsync(() => actionTask);
+            LatestExecution = new AsyncProperty(actionTask);
```
[the remark this answers is not in the document]
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs]: A command may be invoked multiple times. I didn't want to keep track of every invocation (that would basically be a memory leak). Keeping track of the running invocations seemed complicated, especially when the view model will often bock the user from invoking the command again until it finishes. Instead, I just keep track of the last invocation to start, which lets me await the execution in unit tests.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs (hunk 1 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/ProtectedAsyncCommand.cs]: That's how my code generator wrote it. Changing to not be virtual.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/CloudExplorerViewModel.cs (hunk 2 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/CloudExplorerViewModel.cs]: Just readability. I like to only use `var` when the type is obvious (like `var o = new object()`). It is also part of our style guide.

**Pull request merged**